### PR TITLE
Fix for JENKINS-56383 without rolling back fix for JENKINS-38339

### PR DESCRIPTION
**Pull request by sophistifunk**
# Description

Due to some shortcomings in our unit tests (and me not doing enough manual testing it seems) the fix for some edge cases where downstream links weren't showing broke rendering of (some) pipelines with parallel branches.

We rolled back the change as a hotfix (v 1.13.2) as the regression was a more egregious bug than the missing downstream links, but this patch builds on the original fix for 38339 to address the regression.

This patch includes a bunch of new regression tests that explicitly check every node and connection in the graphs, as well as checking for downstream actions where they should be. And a 3 line fix for the regression :) 

See [JENKINS-56383](https://issues.jenkins-ci.org/browse/JENKINS-56383).
See [JENKINS-38339](https://issues.jenkins-ci.org/browse/JENKINS-38339).

# Submitter checklist
- [X] Link to JIRA ticket in description, if appropriate.
- [X] Change is code complete and matches issue description
- [X] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by sophistifunk** [comment]
Every time I say "regression", drink!

/me falls off chair

**Comment by sophistifunk** [comment]
ATH failure in `RestartStageTest` *might* be real, I'm looking into it while I wait for a rebuild.

**Comment by halkeye** [comment]
If it's saying 500 error in the JS logs, I think it's the same issue I've
been poking at since Friday. Nicu said it used to show up sometimes on Linux

**Comment by sophistifunk** [comment]
Never mind!

supergreen

File: blueocean-pipeline-api-impl/src/test/resources/io/jenkins/blueocean/rest/impl/pipeline/restartStage.jenkinsfile [hunk 1]
```diff
@@ -41,7 +41,7 @@ pipeline {
     }
     stage('Deploy') {
       parallel {
-        stage('Deploy') {
+        stage('DeployX') {
```
Comment (halkeye) [inline blueocean-pipeline-api-impl/src/test/resources/io/jenkins/blueocean/rest/impl/pipeline/restartStage.jenkinsfile]: What's this change about?

Comment (sophistifunk) [reply blueocean-pipeline-api-impl/src/test/resources/io/jenkins/blueocean/rest/impl/pipeline/restartStage.jenkinsfile]: The new unit tests are based on stage names for readability, so they won't work on pipelines that have multiple stages with the same name.

**Comment by sophistifunk** [comment]
@halkeye just using existing jenkinsfiles from existing tests. Those tests are looking for different and more specific things, whereas the new tests are explicitly just for ensuring the graph structure. I didn't include any test pipelines that don't finish, or do weird / duplicate stuff.

**Comment by halkeye** [comment]
Then yea looks good to me. Was curious
